src/core/internal/hash.d

```diff
@@ -256,21 +257,21 @@ nothrow pure @safe unittest // issue 19005
 
 //delegate hash. CTFE unsupported
 @trusted @nogc nothrow pure
-size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && is(T == delegate))
+size_t hashOf(T)(scope const T val, size_t seed = 0) if (!is(T == enum) && is(T == delegate))
 {
     assert(!__ctfe, "unable to compute hash of "~T.stringof);
     const(ubyte)[] bytes = (cast(const(ubyte)*)&val)[0 .. T.sizeof];
     return bytesHashAlignedBy!T(bytes, seed);
 }
 
 //class or interface hash. CTFE depends on toHash
-size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && is(T == interface) || is(T == class))
+size_t hashOf(T)(T val, size_t seed = 0) if (!is(T == enum) && is(T == interface) || is(T == class))
```

Comment:
Why not `scope const T`? Even if the `T`'s `toHash` is not qualified appropriately it doesn't matter since we're casting to `Object`. At most we'll need a trusted cast.

Reply:
This whole function should probably be split up into multiple pieces. Note that this will be called if the class is a C++ class, in which case the cast to Object will return null and cause a segfault.